node/src/components/console/tasks.rs

```diff
@@ -0,0 +1,335 @@
+
+        Ok(())
+    }
+}
+
+/// Handler for client connection.
+///
+/// The core loop for the console; reads commands via unix socket and processes them accordingly.
+///
+/// # Security
+///
+/// The handler itself will buffer an unlimited amount of data if no newline is encountered in the
+/// input stream. For this reason ensure that only trusted client connect to the socket producing
+/// the passed in `stream`.
+async fn handler<REv>(
+    effect_builder: EffectBuilder<REv>,
+    stream: UnixStream,
+    mut shutdown_receiver: watch::Receiver<()>,
+) -> io::Result<()>
+where
+    REv: From<DumpConsensusStateRequest> + Send,
+{
+    debug!("accepted new connection on console socket");
+
+    let (reader, mut writer) = stream.into_split();
+    let mut lines = BufReader::new(reader).lines();
+    let mut session = Session::default();
+
+    loop {
+        let shutdown_messages = async { while shutdown_receiver.changed().await.is_ok() {} };
+
+        match future::select(Box::pin(shutdown_messages), Box::pin(lines.next_line())).await {
+            Either::Left(_) => {
+                info!("shutting down console connection to client");
+                break Ok(());
+            }
+            Either::Right((line_result, _)) => {
+                if let Some(line) = line_result? {
+                    session
+                        .process_line(effect_builder, &mut writer, line.as_str())
+                        .await?;
+                } else {
+                    info!("client closed console connection");
+                    break Ok(());
+                }
+            }
+        }
+    }
+}
+
+/// Server task for console.
+pub(super) async fn server<REv>(
+    effect_builder: EffectBuilder<REv>,
+    socket_path: PathBuf,
+    listener: UnixListener,
+    mut shutdown_receiver: watch::Receiver<()>,
+) where
+    REv: From<DumpConsensusStateRequest> + Send,
+{
+    let handling_shutdown_receiver = shutdown_receiver.clone();
+    let mut next_client_id: u64 = 0;
+    let accept_connections = async move {
+        loop {
+            match listener.accept().await {
+                Ok((stream, client_addr)) => {
+                    let client_id = next_client_id;
+
+                    let span = info_span!("console", client_id,);
+
+                    span.in_scope(|| {
+                        info!(client_addr = %ShowUnixAddr(&client_addr), "accepted connection");
+                    });
+
+                    next_client_id += 1;
+
+                    tokio::spawn(
+                        handler(effect_builder, stream, handling_shutdown_receiver.clone())
+                            .instrument(span),
+                    );
+                }
+                Err(err) => {
+                    info!(%err, "failed to accept incoming connection on console socket");
+                }
+            }
+        }
+    };
+
+    let shutdown_messages = async move { while shutdown_receiver.changed().await.is_ok() {} };
```

Comment:
Just checking my understanding: this future will run until `shutdown_receiver.changed()` returns `Err`? Shouldn't we consider any signal sent on to that receiver as a shutdown signal?

Reply:
We never send anything down the receiver, `Err` means the other end was dropped, which is the signal to shutdown.

Reply:
I see! I think I thought that when a sender is dropped it will change the shared value. Thanks for explanation.